### ISSUE TITLE
ld: library not found (library search path's hash doesn't match rugby cache's hash)

### ISSUE THREAD

**Issue by PRESIDENT810** [swiftyfinch/Rugby#256]
## Describe the bug
I was integrating rugby into our CI builds so we can utilize the pod cache and make our builds faster.

However I run into some issues. Not frequent but hard to reproduce. This is an error from one of our CI build:
<img width="1287" alt="image" src="https://github.com/swiftyfinch/Rugby/assets/44538064/c1de2af7-743d-4497-a22f-9fffd36c4e5d">

When it was building our iOS app, it aws trying to find RNReanimated library in `~/.rugby/bin/RNReanimated/Internal-iphoneos-arm64/5e5e9f6` path, and it failed. I sshed to that machine and find the only RNReanimated cache directory is `~/.rugby/bin/RNReanimated/Internal-iphoneos-arm64/1dbf35c/`.

I don't have any idea why the hash of the actual cached library's path is different with the search path. Also, when I retry the build on the same machine, it worked with the correct path.

I find this is hard to reproduce, since on CI every build clears the build directory and it's difficult to preserve the crime scene. I tried to reproduce this locally but I also can't trigger this on local builds.

I wonder why the search path's hash is different with the cache's hash? I suppose they should be generated from the same hashing mechanism?

Version: 2.1.1

**Comment by swiftyfinch** [comment]
Hello @PRESIDENT810!

Could you attach both hash files? Or you can use `diff 1dbf35c.txt 5e5e9f6.txt` by yourself.
Usually there are two issues:
1) Do you use the same Xcode version on CI? In your file `1dbf35c.txt` the last line is `swift_version: 5.7.2`. So, we need to be sure that in `5e5e9f6.txt` it will be the same.
2) Do you have some scripts in your build phases which can edit source code during building? If so, you should run all of them before Rugby. Otherwise, after build is finished the hashes will be different, because it's different source code.

**Comment by PRESIDENT810** [comment]
Hey @swiftyfinch 
The problem is I don't have 5e5e9f6.yml dumped by rugby.
<img width="493" alt="image" src="https://github.com/swiftyfinch/Rugby/assets/44538064/aa374955-cc44-4534-bdb8-9cbca7466747">
As you can see the only cached pod is 1dbf35c. I was think of dumping two yaml file and compare the difference, too, but I didn't expect not seeing 5e5e9f6 under rugby cache directory.

I'm also pretty certain that I'm using the same Xcode version:
<img width="1305" alt="image" src="https://github.com/swiftyfinch/Rugby/assets/44538064/bcffa986-b019-4c9d-a848-f29a42bfa50f">

<img width="1317" alt="image" src="https://github.com/swiftyfinch/Rugby/assets/44538064/fde4918d-10d7-451c-906a-7a1fc30adc6e">

There is also no other steps between rugby cache and iOS builds:
<img width="1278" alt="image" src="https://github.com/swiftyfinch/Rugby/assets/44538064/045848f9-011c-4031-8855-de7196eda16e">

However we do have multiple iOS builds on a same machine (both GitHub workflow and Jenkins run on the same host but different workspace, and GitHub workflow is using rugby cache while Jenkins is not), could this be a problem?

**Comment by swiftyfinch** [comment]
@PRESIDENT810 Oh, I noticed that you have absolute paths in your hash file.
```yml
- buildActionMask: 2147483647
  files:
  - '/Users/local-mobile/actions-runner/_work/mobile-app/mobile-app/node_modules/react-native-reanimated/Common/cpp/AnimatedSensor/AnimatedSensorModule.h: f2a5e39'
  - '/Users/local-mobile/actions-runner/_work/mobile-app/mobile-app/node_modules/react-native-reanimated/Common/cpp/LayoutAnimations/LayoutAnimationsProxy.h: c898f4c'
  - '/Users/local-mobile/actions-runner/_work/mobile-app/mobile-app/node_modules/react-native-reanimated/Common/cpp/NativeModules/NativeReanimatedModule.h: 81904ae'
```

It's strange and should be like this:
```yml
- buildActionMask: 2147483647
  files:
  - 'Pods/Alamofire/Source/AFError.swift: dd80e59'
  - 'Pods/Alamofire/Source/Alamofire.swift: f5b8707'
  - 'Pods/Alamofire/Source/AlamofireExtended.swift: 6b508cd'
  - 'Pods/Alamofire/Source/AuthenticationInterceptor.swift: 8b97282'
```

So, I think it's kind of bug. Rugby cuts paths relative to `Pods` directory, but you have files beyond it.
And the problem is I haven't idea how to make stable path for any machine outside of `Pods` directory. We can add project name to it, but the same binaries can't be reused between different projects.

Maybe we can try to cut relative to `Pods` at first, and if it fails, we can try again to cut relative parent directory.

**Comment by PRESIDENT810** [comment]
Hmm I'm not an iOS expert but I guess this has something to do with using react native? Like some source files from RN are not in `Pods` directory so things might go wrong?

I could exclude those RN related Pods as a temporary workaround, but I still wonder why this is not 100% reproducible. I tried locally and the path is also absolute in the dumped YAML file, yet I never managed to reproduce this on local.

**Comment by PRESIDENT810** [comment]
HA I just found out how to reproduce this. It turns out that this is because of a React Native generated file under build/generated/ios directory.
```
.
├── FBReactNativeSpec
│   ├── FBReactNativeSpec-generated.mm
│   └── FBReactNativeSpec.h
└── React-Codegen.podspec.json
```
This is how this directory look like, and in `React-Codegen.podspec.json`:
```
{
    "name": "React-Codegen",
    "version": "0.71.8",
    "summary": "Temp pod for generated files for React Native",
    "homepage": "https://facebook.com/",
    "license": "Unlicense",
    "authors": "Facebook",
    "compiler_flags": "-DFOLLY_NO_CONFIG -DFOLLY_MOBILE=1 -DFOLLY_USE_LIBCPP=1 -Wno-comma -Wno-shorten-64-to-32 -Wno-documentation -Wno-nullability-completeness -std=c++17",
    "source": {
        "git": ""
    },
    "header_mappings_dir": "./",
    "platforms": {
        "ios": "12.4"
    },
    "source_files": "**/*.{h,mm,cpp}",
    "pod_target_xcconfig": {
        "HEADER_SEARCH_PATHS": "\"$(PODS_ROOT)/boost\" \"$(PODS_ROOT)/RCT-Folly\" \"${PODS_ROOT}/Headers/Public/React-Codegen/react/renderer/components\" \"$(PODS_ROOT)/Headers/Private/React-Fabric\" \"$(PODS_ROOT)/Headers/Private/React-RCTFabric\""
    },
    "dependencies": {
        "FBReactNativeSpec": [],
        "React-jsiexecutor": [],
        "RCT-Folly": [],
        "RCTRequired": [],
        "RCTTypeSafety": [],
        "React-Core": [],
        "React-jsi": [],
        "ReactCommon/turbomodule/bridging": [],
        "ReactCommon/turbomodule/core": [],
        "hermes-engine": []
    }
}
```

Before `rugby build`, both `FBReactNativeSpec/FBReactNativeSpec-generated.mm` and `FBReactNativeSpec/FBReactNativeSpec.h` are empty, but after `rugby build` there will be RN generated code in these files.
The comment says "This code was generated by [react-native-codegen](https://www.npmjs.com/package/react-native-codegen)."

Then when do `rugby use`, the content in those files are changed, which leads to a different hash, which affects pods depend on this pod, and so we can't find the library.

I'm not very familiar with RN so I'm not sure what cause RN generates such code when building Pods. I guess the reason why this is hard to reproduce is because you have to clear all the generated stuff, otherwise once the source is generated it's there, and subsequent builds will use the generated source file, which hides the problem.

**Comment by swiftyfinch** [comment]
@PRESIDENT810 Yeah, I see. I think there is an build phase script in a target of your project. Something like `React Native Codegen`.

I think that we can add the ability to run all build phase scripts above `Compile Sources` during warmup command. It will be pretty convenient for all users.

So, I think that we are ready to sum up:
1. We need to add fallback scenario to resolve paths relative to parent directory, if we can't do it relative to the Pods directory;
2. We need to add ability to run all build phase scripts above `Compile Sources` as part of `rugby warmup` command.

Am I right? Do you want to add something?
Also, I don't have any experience in React Native. I wonder if you could provide a light project to test. It'll be much helpful for me. Or I need to read something about how to install React Native and prepare a test project.

**Comment by PRESIDENT810** [comment]
@swiftyfinch 
I don't quite sure how this works, but following the React Native tutorial, I managed to reproduce this in a demo project: https://github.com/PRESIDENT810/RugbyTest. Instructions to reproduce is attached in README.

I thought React Native Codegen could be generating code, but I can't find where it did so. This target looks like just like other normal targets, and there is no shell script is executed or other suspicious steps, unlike "Copy xcframeworks" or "Generate headers" I mentioned in another issue:
<img width="1645" alt="image" src="https://github.com/swiftyfinch/Rugby/assets/44538064/4cc50347-33db-42c7-a4ee-a728dac04d2d">

Hope this helps!

**Comment by swiftyfinch** [comment]
@PRESIDENT810 Thank you! I will try 🚀

**Comment by PRESIDENT810** [comment]
@swiftyfinch I think I know what's going on. This is more or less similar to the previous issue I mentioned in #250 

So when you are using React Native, you should need to build this React-Codegen library, which has the following dependencies:
<img width="703" alt="image" src="https://github.com/swiftyfinch/Rugby/assets/44538064/07f4a39a-86bd-4e26-aed9-620e833465f4">
This FBReactNativeSpec is a target running some script
<img width="1574" alt="image" src="https://github.com/swiftyfinch/Rugby/assets/44538064/17affcc8-b58d-4b2e-bcd7-5ebfb41731c8">
And this script will generate code that makes hash change.

What I'm thinking (may not be correct):
1. We need to generate hash when all source files are finalized, including modification done by some pods and their dependencies. In this case, React-Codegen depends on FBReactNativeSpec, which modifies source code. If user doesn't know this, they will falsely think all source files are finalized after pod install, whereas in this case, source files are only finalized after FBReactNativeSpec get built.
2. We might need to inspect targets by topology sorted order, and then inspect their source files to determine if cache can be reused. In this case, every time FBReactNativeSpec generates the same code, so after the first build this is idempotent for subsequent builds, which is fine as long as you don't touch the generated source code. However, if someone has a target A has a dependency B, which relies on some external inputs that change frequently, like code generated from protobuf schema via a shell script, **we need to first build B before we compute the hash based on source files**, and then decide if we can reuse A's cache. If we just look at A's source code without building B, we may not be able to capture the change brought by pb schema change. Probably we should first build all non-cached targets, and then let rugby determine if the cached targets can be reused?

I guess the simplest way to work around this is just manually check every target and exclude suspicious ones. I doubt if xcodebuild can be as smart as bazel or gradle to discover subtle changes like this, since Xcode doesn't know what the input and output is when executing shell scripts.

**Comment by swiftyfinch** [comment]
@PRESIDENT810 Yeah, I think you are right! Thanks for an additional information!

Yesterday, I found the same build phase `FBReactNativeSpec`.
I'm not sure that we need to traverse in the right order during pre-build phase.
I really hope that on this stage all modules are independent.

So, I'm going to add just an additional step in the build and the warmup commands.
This step should find all targets with build phase script which placed above source files phase. And then we can try to run these scripts passing the right environment variables into them. I think we can do it in parallel without any problem. But will see)

I've already written similar solution, but in Ruby script outside of Rugby.
Now, I see that you need it and it will be useful for all Rugby users.
It will be nice to have such step in Rugby natively.

**Comment by PRESIDENT810** [comment]
@swiftyfinch Thanks! I guess that will be helpful (and even more helpful for RN users)

Yet I'm not sure if running scripts in parallel can be safe. I guess for some extreme edge cases where one script depends on the output of another script, we can't simply run them in parallel. Rather we need to construct the DAG and run these scripts according to the order of their dependencies. I think we can run scripts without building the targets they belong as long as they don't rely on built artifacts (I guess these are just rare cases)?

I think it's also helpful to warn people like "You should be cautious to cache targets that you are not certain if their input will not be affected by other targets' output"  so they will know there could be potential tricky scenarios like this, and only cache those relatively independent targets.

**Comment by swiftyfinch** [comment]
@PRESIDENT810 Yeah, it's all about time. I think it will be more faster to run it in parallel.
But, if it breaks something we can add option to run it in the right order.

**Comment by swiftyfinch** [comment]
@PRESIDENT810 Hello! Could you test this one #263?
I've tested on my projects ✅.

**Comment by PRESIDENT810** [comment]
@swiftyfinch Unfortunately there is still one problem...

Our project is using EXConstants, which I guess is from expo. So we have two targets
- The first target is EXConstants-EXConstants, which doesn't do anything but merely put `Info.plist` into `EXConstants.bundle` under `~/Library/Developer/Xcode/DerivedData/<PROJ_DIR>/Build/Products/Internal-iphoneos/EXConstants/EXConstants.bundle`
- The second target is EXConstants, which has a step "[CP-User] Generate app.config for prebuilt Constants.manifest". This step executes a shell script and generates `app.config` inside that `EXConstants.bundle`. The content of this script is:
```
#!/usr/bin/env bash

set -eo pipefail

DEST="$CONFIGURATION_BUILD_DIR"
RESOURCE_BUNDLE_NAME="EXConstants.bundle"
EXPO_CONSTANTS_PACKAGE_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd -P)"

# For classic main project build phases integration, will be no-op to prevent duplicated app.config creation.
#
# `$PROJECT_DIR` is passed by Xcode as the directory to the xcodeproj file.
# in classic main project setup it is something like /path/to/app/ios
# in new style pod project setup it is something like /path/to/app/ios/Pods
PROJECT_DIR_BASENAME=$(basename $PROJECT_DIR)
if [ "x$PROJECT_DIR_BASENAME" != "xPods" ]; then
  exit 0
fi

# If PROJECT_ROOT is not specified, fallback to use Xcode PROJECT_DIR
PROJECT_ROOT=${PROJECT_ROOT:-"$PROJECT_DIR/../.."}
PROJECT_ROOT=${PROJECT_ROOT:-"$EXPO_CONSTANTS_PACKAGE_DIR/../.."}

cd "$PROJECT_ROOT" || exit

"${EXPO_CONSTANTS_PACKAGE_DIR}/scripts/with-node.sh" "${EXPO_CONSTANTS_PACKAGE_DIR}/scripts/getAppConfig.js" "$PROJECT_ROOT" "$DEST/$RESOURCE_BUNDLE_NAME"
```
As this script is executed in the prebuild stage, `CONFIGURATION_BUILD_DIR` will point at `.rugby/build/Internal-iphoneos`, but this is not where `EXConstants.bundle` is. This causes this error:
```
⛔️ Error: Build failed.
 ✖︎ Error: ENOENT: no such file or directory, open '.rugby/build/Internal-iphoneos/EXConstants/EXConstants.bundle/app.config'
```
This is so hard, I guess Rugby just doesn't like React Native...

**Comment by swiftyfinch** [comment]
@PRESIDENT810 Could you add this issue to your RugbyTest repository?

It sounds strange to me, because I think that if the issue depends on `$CONFIGURATION_BUILD_DIR` we should have a lot of reports. Do you see this issue if you build your project with 🏈 Rugby twice without `--prebuild`?

Also, could you share rugby commands which you call?
Are you sure that you pass the same arguments to `build pre` and `build full`?

**Comment by PRESIDENT810** [comment]
@swiftyfinch Sure I'm trying to reproduce this... I don't have much mobile dev experience so I'm not really sure how this works either. Still working on it and will let you know once I know how to do it.

And sorry my previous description is wrong. I thought rugby is building `EXConstants.bundle` in a wrong location (`~/Library/Developer/Xcode/DerivedData/<PROJ_DIR>/Build/Products/Internal-iphoneos/EXConstants/`), but actually I guess this directory is generated in previous xcode builds (builds without rugby). I mistankenly thought that rugby is looking for `EXConstants.bundle` using a wrong path.

After I deleted `DerivedData` and `~/.rugby/bin`, I no longer see `~/Library/Developer/Xcode/DerivedData/<PROJ_DIR>/Build/Products/Internal-iphoneos/EXConstants/`. I think what really happened is during `rugby build pre` stage, since target `EXConstants-EXConstants` is a target without any scripts, prebuild stage simply skips this target. Then it finds target `EXConstants`, which depends on `EXConstants-EXConstants` and its script generates `app.config` in `EXConstants.bundle`. However, since we haven't reached build stage yet, `EXConstants-EXConstants` is not built so `EXConstants.bundle` is not created yet.

I feel weird why we have `EXConstants-EXConstants` too. It's just an empty target that only creates a bundle:
<img width="1231" alt="image" src="https://github.com/swiftyfinch/Rugby/assets/44538064/7cc5c95a-912c-4d1b-8979-6a1642a80540">

**Comment by swiftyfinch** [comment]
@PRESIDENT810 I've updated the pull request.
Could you test it again?

Please, try two different calls:
```
rugby cache --prebuild
```
```
RUGBY_DELETE_PRE_TARGETS=YES rugby cache --prebuild
```
I think that both should work. But how much time the `Prebuild` step will take?

**Comment by PRESIDENT810** [comment]
@swiftyfinch Thanks, that works!

As for the time:
```
⚑ Cache --prebuild --config Internal --sdk ios --arch arm64 --except react-native-config
  ⚑ Prebuild
    ✓ CLT: Xcode 14.2
    ✓ Finding Build Targets
    ✓ Removing Build Phases
    ✓ [0.2s] Creating Build Target
    ✓ [0.4s] Saving Project
    ✓ [11.7s] Internal: ios-arm64 (176)
  ✓ [13.5s] Prebuild
  ⚑ Build
    ✓ CLT: Xcode 14.2
    ✓ Finding Build Targets
    ✓ [0.1s] Checking Binaries Storage
    ✓ [1.5s] Hashing Targets
    ✓ Finding Binaries
    ✓ Found 99% Binaries (158/159)
    ⚑ Reusing Binaries
      ✓ Patching Product Files
      ✓ Deleting Targets (158)
    ✓ [0.1s] Reusing Binaries
    ✓ [0.1s] Saving Project
    ✓ [0.3s] Creating Build Target
    ✓ [0.1s] Saving Project
    ✓ [12.6s] Internal: ios-arm64 (1)
    ✓ Saving binaries (1)
  ✓ [16.6s] Build
  ⚑ Use
    ✓ Finding Build Targets
    ✓ [1.2s] Hashing Targets
    ✓ Patching Product Files
    ✓ Deleting Targets (159)
    ✓ [0.1s] Saving Project
  ✓ [2.7s] Use
✓ [32.9s] Cache
⚑ Let's Roll-oll 🏈
```

**Comment by swiftyfinch** [comment]
@PRESIDENT810 Great news! 🎉

But I wonder, did you try both calls?
You sent only one output.

Did you try to set the environment varialbe `RUGBY_DELETE_PRE_TARGETS=YES`?
```sh
RUGBY_DELETE_PRE_TARGETS=YES rugby cache --prebuild --config Internal --sdk ios --arch arm64 --except react-native-config
```

**Comment by swiftyfinch** [comment]
> @PRESIDENT810 Great news! 🎉
> 
> But I wonder, did you try both calls? You sent only one output.
> 
> Did you try to set the environment varialbe `RUGBY_DELETE_PRE_TARGETS=YES`?
> 
> ```shell
> RUGBY_DELETE_PRE_TARGETS=YES rugby cache --prebuild --config Internal --sdk ios --arch arm64 --except react-native-config
> ```

Oh, please forget about `RUGBY_DELETE_PRE_TARGETS` 😅
But could you test it again? I hope that I've made the final changes.

**Comment by PRESIDENT810** [comment]
@swiftyfinch Ah sorry! I tried both commands and they both worked, so I thought they were the same...
Result for `rugby cache --prebuild`:
```
⚑ Cache --prebuild --config Internal --sdk ios --arch arm64 --except react-native-config
  ⚑ Prebuild
    ✓ CLT: Xcode 14.2
    ✓ Finding Build Targets
    ✓ Removing Build Phases
    ✓ Deleting Targets
    ✓ Creating Build Target
    ✓ [0.1s] Saving Project
    ✓ [3s] Internal: ios-arm64 (1)
  ✓ [4.9s] Prebuild
  ⚑ Build
    ✓ CLT: Xcode 14.2
    ✓ Finding Build Targets
    ✓ [0.1s] Checking Binaries Storage
    ✓ [1.2s] Hashing Targets
    ✓ Finding Binaries
    ✓ Found 100% Binaries (159)
  ✓ [2.8s] Build
  ⚑ Use
    ✓ Finding Build Targets
    ✓ Hashing Targets
    ✓ Patching Product Files
    ✓ Deleting Targets (159)
    ✓ [0.1s] Saving Project
  ✓ [0.7s] Use
✓ [8.5s] Cache
⚑ Let's Roll-oll 🏈
```
Result for `RUGBY_DELETE_PRE_TARGETS=YES rugby cache --prebuild`:
```
⚑ Cache --prebuild --config Internal --sdk ios --arch arm64 --except react-native-config
  ⚑ Prebuild
    ✓ CLT: Xcode 14.2
    ✓ Finding Build Targets
    ✓ Removing Build Phases
    ✓ [0.1s] Deleting Targets
    ✓ Creating Build Target
    ✓ [0.2s] Saving Project
    ✓ [2.9s] Internal: ios-arm64 (1)
  ✓ [4.7s] Prebuild
  ⚑ Build
    ✓ CLT: Xcode 14.2
    ✓ Finding Build Targets
    ✓ [0.1s] Checking Binaries Storage
    ✓ [1.2s] Hashing Targets
    ✓ Finding Binaries
    ✓ Found 100% Binaries (159)
  ✓ [2.8s] Build
  ⚑ Use
    ✓ Finding Build Targets
    ✓ Hashing Targets
    ✓ Patching Product Files
    ✓ Deleting Targets (159)
    ✓ [0.1s] Saving Project
  ✓ [0.4s] Use
✓ [8.1s] Cache
⚑ Let's Roll-oll 🏈
```

**Comment by swiftyfinch** [comment]
@PRESIDENT810 Thank you! I've already removed `RUGBY_DELETE_PRE_TARGETS` 😅
OK, I think we are ready to merge this 🚀

**Comment by swiftyfinch** [comment]
It's finally done in #263
@PRESIDENT810 Thank you for your contribution! 🚀
